File: pet/macos/Sources/SemaPet/Protocol.swift
import Foundation

enum PetState: String, Codable {
    case idle, thinking, working, attention, sleeping

    var priority: Int {
        switch self {
        case .attention: return 4
        case .working:   return 3
        case .thinking:  return 2
        case .idle:      return 1
        case .sleeping:  return 0
        }
    }
}

struct RegisterPayload: Codable {
    let sessionId: String
    let cwd: String
}

struct UnregisterPayload: Codable {
    let sessionId: String
}

struct StatePayload: Codable {
    let sessionId: String
    let state: PetState
    let ts: Double?
}

struct SayPayload: Codable {
    let sessionId: String
    let text: String
    let kind: String?
    let ttlMs: Double?
    let sticky: Bool?
}

struct FocusCommand: Codable {
    let type: String
    let sessionId: String
}

struct NoopCommand: Codable {
    let type: String
}

enum PetConstants {
    static let host = "127.0.0.1"
    static let port: UInt16 = 24700
    static let serverHeaderKey = "x-sema-pet"
    static let serverHeaderValue = "server"
    static let longPollTimeoutSec: TimeInterval = 25
    static let idleToSleepingSec: TimeInterval = 60 * 60
    static let sayDefaultTtlMs: Double = 5000
    static let sayMaxChars: Int = 40
}

File: pet/windows/src/protocol.rs
use serde::{Deserialize, Serialize};

use crate::bubble_store::BubbleKind;

pub const PET_HOST: &str = "127.0.0.1";
pub const PET_PORT: u16 = 24700;
pub const PET_SERVER_HEADER: &str = "x-sema-pet";
pub const PET_SERVER_HEADER_VALUE: &str = "server";
pub const SAY_DEFAULT_TTL_MS: u64 = 5000;
pub const SAY_MAX_VISIBLE: usize = 3;
pub const SAY_MAX_CHARS: usize = 40;

#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Deserialize, Serialize)]
#[serde(rename_all = "lowercase")]
pub enum PetState {
    Idle,
    Thinking,
    Working,
    Attention,
    Sleeping,
}

impl PetState {
    pub fn priority(self) -> u8 {
        match self {
            PetState::Attention => 4,
            PetState::Working => 3,
            PetState::Thinking => 2,
            PetState::Idle => 1,
            PetState::Sleeping => 0,
        }
    }
}

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct RegisterPayload {
    pub session_id: String,
    pub cwd: String,
    pub client_pid: Option<u32>,
}

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct UnregisterPayload {
    pub session_id: String,
}

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct StatePayload {
    pub session_id: String,
    pub state: PetState,
    pub ts: u64,
}

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SayPayload {
    pub session_id: String,
    pub text: String,
    pub kind: Option<BubbleKind>,
    pub ttl_ms: Option<u64>,
    pub sticky: Option<bool>,
}

#[derive(Debug, Serialize, PartialEq, Eq)]
#[serde(
    tag = "type",
    rename_all = "lowercase",
    rename_all_fields = "camelCase"
)]
pub enum PetCommand {
    Focus { session_id: String },
    Noop,
}

#[derive(Debug, Serialize)]
pub struct HealthResponse {
    pub ok: bool,
}
